### ISSUE TITLE
When hex_erl is included as project dependency it tries to run protobuf provider hooks

### ISSUE THREAD

**Issue by kubaodias** [hexpm/hex_erl#20]
I have a project that includes hex_erl in rebar.config e.g.
```
{deps,[
    {hex_erl, {git, "https://github.com/hexpm/hex_erl.git", {branch, "master"}}}
]}.
```
When I run ```rebar3 compile``` in my project I get the following error:
```
===> Verifying dependencies...
===> Fetching hex_erl ({git,"https://github.com/hexpm/hex_erl.git",
                                  {ref,
                                      "6c8973c2ce3acea1ebec8f08d23b4951f272361b"}})
===> Compiling hex_erl
===> Unable to run pre hooks for 'compile', command 'compile' in namespace 'protobuf' not found.
```
It seems that when protobuf is added as project plugin (https://github.com/hexpm/hex_erl/blob/6c8973c2ce3acea1ebec8f08d23b4951f272361b/rebar.config#L5) it isn't fetched as project dependency. Please refer to end of this section in rebar3 documentation: http://www.rebar3.org/v3/docs/using-available-plugins#section-project-plugins-and-overriding-commands

**Comment by ericmj** [comment]
If I understand correctly rebar3_gpb_plugin should be moved to `plugins` and rebar3_proper kept in `project_plugins`. Please let us know if making this change fixes the problem for you.

**Comment by ericmj** [comment]
Alternatively we can remove protobuf from `provider_hooks` and manually run the protobuf generator, I think this would be preferable.

**Comment by tsloughter** [comment]
No, `rebar3_proper` should definitely be a project plugin. There is no reason to have it get downloaded when the project is being used as a dependency. That is one of the purposes of project plugins.

As for gpb, unless the generated modules are committed then yes it needs to be a regular plugin.

**Comment by wojtekmach** [comment]
@tsloughter on `master` both plugins are project plugins, and yet the keep being downloaded when fetching the package, is there a way to avoid downloads?

See: https://github.com/hexpm/hex_erl/pull/23/files - it's safe to remove provider hooks because we commit generated .erl from .pb. On that branch I switch project_plugins to plugins, btw, which I guess I'll revert, but in either case stuff is being downloaded. Help would be very appreciated.

**Comment by tsloughter** [comment]
It is a bug in rebar3 if a project plugin is fetched when the application is used as a dependency of another project.

**Comment by wojtekmach** [comment]
what's weird is when I add hex_erl on a rebar3 project and do `rebar3 compile` it *does not* download gbp, but if I do it on a new mix project it does, even if I switch my ~/.mix/rebar3 version to latest. If someone can reproduce it on mix I'd appreciate, maybe the issue is there after all?

For now, I've removed provider hooks and so part of the issue is solved.

**Comment by ferd** [comment]
Mix downloads deps its own way, and only gives paths for rebar3 to compile.

**Comment by ericmj** [comment]
@ferd Mix downloads deps but this is a plugin so it would be rebar3 downloading it when running the bare compiler.

**Comment by ericmj** [comment]
My guess is that rebar3 bare compiler runs `project_plugins` because it doesn't know it is being used a dependency. If that's the case the bare compiler should change to behave as a dependency compilation instead or a flag be added that tells it to do so.

**Comment by ferd** [comment]
ah that could be a possible interpretation. We have no way to know if the bare compiler is used for a dep or a top-level app (another tool than mix could wrap it), and maybe a switch should be needed for that.

**Comment by tsloughter** [comment]
Yea, that'd be the issue. Not sure if it is easily changed.
